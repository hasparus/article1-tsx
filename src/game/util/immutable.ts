import { IntToString } from './IntToString';

type Updated<T, I extends string, X> = { 
  [K in keyof T]: K extends I ? X : T[K]
}

/**
 * Replace ith element of xs with x
 */
export function update<
  Xs extends readonly any[],
  I extends number,
  X
>(
  xs: Xs,
  i: I,
  x: number extends I ? Xs[number] : X
): number extends I ? Xs : Updated<Xs, IntToString<I>, X> {
  console.assert(i < xs.length, "i should be less than xs.length");
  return [...xs.slice(0, i), x, ...xs.slice(i + 1)] as any;
}
