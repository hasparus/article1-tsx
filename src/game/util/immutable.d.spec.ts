import { update } from "./immutable";

const abc = ["a", "b", "c"] as ["a", "b", "c"];
const kh = "kh" as const;

// @dts-jest:pass
const _t1: readonly ["a", "b", "kh"] = update(abc, 2, kh);

// @dts-jest:pass
const _t2: ["a", "b", "kh"] = update(abc, 2, kh);

// @dts-jest:pass
const _t3: [5.20, 7.34] = update([5.20, 7.34], 1, 7.34 as const);

// @dts-jest:pass
const _t4: [5.20, 7.34] = update([5.20, 1.34], 1, 7.34 as const);

const one: number = 1;

// @dts-jest:pass
const _t5: ['a', 'b'] = update(['a', 'b'], one, 'a')

// @dts-jest:fail:snap
const _t6: ['a', 'b'] = update(['a', 'b'], one, 'c')

// @dts-jest:pass
const _t7 = update(['a', 'b'], one, 'c') // => string
