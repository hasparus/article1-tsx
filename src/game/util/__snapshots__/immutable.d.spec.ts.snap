// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`const _t6: ['a', 'b'] = update(['a', 'b'], one, 'c') (type) should match snapshot: const _t6: ['a', 'b'] = update(['a', 'b'], one, 'c') 1`] = `"Argument of type '\\"c\\"' is not assignable to parameter of type '\\"a\\" | \\"b\\"'."`;
